### PR TITLE
Fix and tests for tree node modifications prior to expansion

### LINKED ISSUE #2432: Modifying Tree Widget nodes before the Tree is Expanded/Drawn in macOS
### Describe the bug

If you attempt to append, insert, remove, or modify a child node used as a data source for the Tree widget, before it has expanded and displayed those nodes, an **AttributeError: 'Node' object has no attribute '_impl'** will be generated. It appears that the child nodes are being generated on demand and don't exist until they have been displayed. A work around is to manually expand the Tree before doing any modifications.

### Steps to reproduce

A minimum script to reproduce the error is shown below. If you comment out `tree.expand()` the error will be thrown.

```
import toga

def build(app):
    tree = toga.Tree(headings=["Type", "Text", "Options", "ID"])
    root_node = tree.data.append(dict(type="Root Node", text="", options="", ID="0"))
    leaf_node = dict(type="Spelling", text="word", options="opt", id="0", root=False)
    tree.expand()
    #root_node.append(leaf_node)
    root_node.insert(0, leaf_node)
    return tree

def main():
    return toga.App("Node Test App", "au.com.reefwing.node_test", startup=build)

if __name__ == "__main__":
    main().main_loop()
```

### Expected behavior

It should be possible to modify the underlying Tree data model regardless of the state of the UI widget.

### Screenshots

_No response_

### Environment

- Operating System: macOS
- Python version: 3.11
- Software versions:
  - Briefcase:
  - Toga: 0.4.2

### Logs

```
File "/Users/dsuch/Documents/BeeWare/tests/node_test.py", line 14, in build
    root_node.insert(0, leaf_node)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/toga/sources/tree_source.py", line 122, in insert
    self._source.notify("insert", parent=self, index=index, item=node)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/toga/sources/base.py", line 84, in notify
    method(**kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/toga_cocoa/widgets/tree.py", line 223, in insert
    inParent=parent._impl if parent else None,
             ^^^^^^^^^^^^
AttributeError: 'Node' object has no attribute '_impl'
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by reefwing**
Appending/inserting/changing/removing a tree data node prior to its expansion was throwing an AttributeError.  As an optimisation, objects representing tree nodes are created on first render. Trying to access these prior to tree expansion is causing the issue.

Added a node_impl() method that can be used anywhere that the user requests node._impl, and ensures that the impl exists.

Fixes #2432.

## PR Checklist:

- [x ] All new features have been tested
- [x ] All new features have been documented
- [x ] I have read the **CONTRIBUTING.md** file
- [x ] I will abide by the code of conduct

### CODE REVIEW

File: testbed/tests/widgets/test_tree.py [hunk 1]
```diff
@@ -459,6 +540,76 @@ async def _row_change_test(widget, probe):
     probe.assert_cell_content((0, 4), 1, "4")
     probe.assert_cell_content((0, 4), 2, "<data 4>")
 
+    # Test append/insert/modify/remove prior to tree expansion
+    # Append a row to the table
```
Comment (freakboy3742) [inline testbed/tests/widgets/test_tree.py:544]:
```suggestion
    # Append a child to a non-expanded root
```

File: testbed/tests/widgets/test_tree.py [hunk 2]
```diff
@@ -459,6 +540,76 @@ async def _row_change_test(widget, probe):
...
+    widget.data[0].append({"a": "AX", "b": "BX", "c": "CX"})
+    await probe.redraw("Full row has been appended")
```
Comment (freakboy3742) [inline testbed/tests/widgets/test_tree.py:546]:
```suggestion
    await probe.redraw("Full row has been appended to a non-expanded root")
```

File: testbed/tests/widgets/test_tree.py [hunk 3]
```diff
@@ -459,6 +540,76 @@ async def _row_change_test(widget, probe):
...
+
+    assert probe.child_count((0,)) == 6
+    probe.assert_cell_content((0, 4), 0, "A4")
+    probe.assert_cell_content((0, 5), 0, "AX")
+
+    # Insert a row into the middle of the table;
+    # Row is missing a B accessor
+    widget.data[0].insert(2, {"a": "AY", "c": "CY"})
+    await probe.redraw("Partial row has been appended")
```
Comment (freakboy3742) [inline testbed/tests/widgets/test_tree.py:555]:
```suggestion
    await probe.redraw("Partial row has been appended to a non-expanded root")
```

File: testbed/tests/widgets/test_tree.py [hunk 4]
```diff
@@ -459,6 +540,76 @@ async def _row_change_test(widget, probe):
...
+    probe.assert_cell_content((0, 5), 0, "A4")
+    probe.assert_cell_content((0, 6), 0, "AX")
+
+    # Missing value has been populated
+    probe.assert_cell_content((0, 2), 1, "MISSING!")
+
+    # Change content on the partial row
+    widget.data[0][2].a = "ANEW"
+    widget.data[0][2].b = "BNEW"
+    await probe.redraw("Partial row has been updated")
```
Comment (freakboy3742) [inline testbed/tests/widgets/test_tree.py:568]:
```suggestion
    await probe.redraw("Partial non-visible row has been updated")
```

File: testbed/tests/widgets/test_tree.py [hunk 5]
```diff
@@ -459,6 +540,76 @@ async def _row_change_test(widget, probe):
...
+    probe.assert_cell_content((0, 2), 0, "ANEW")
+    probe.assert_cell_content((0, 5), 0, "A4")
+    probe.assert_cell_content((0, 6), 0, "AX")
+
+    # Missing value has the default empty string
+    probe.assert_cell_content((0, 2), 1, "BNEW")
+
+    # Delete a row
+    del widget.data[0][3]
+    await probe.redraw("Row has been removed")
```
Comment (freakboy3742) [inline testbed/tests/widgets/test_tree.py:580]:
```suggestion
    await probe.redraw("Non-visible row has been removed")
```

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This looks really good - nice work!

I've made a couple of suggestions inline; these are mostly minor stylistic changes and some tweaks to some descriptions.

I can see 2 substantive changes that are required - both additional tests.

The first is to exercise the case of adding a child to a hidden child. The tests you've added validate that adding a child to a visible, but non-expanded root work; but you can also add a node to a child that has *never* been visible. The implementation you've provided should work for this case; but it's not explicitly tested.

The test doesn't need to do much - something like: 
```
    widget.data[0][5].append({"a": "AXX", "b": "BXX", "c": "CXX"})
```
after adding the row to the non-expanded root (around L550), plus some subsequent content and child count checks.

The second check is that once you've finished adding nodes to non-visible nodes and non-expanded roots, you can successfully expand the tree and see the result of all your changes.

As an aside - I don't know if this was intentional, but you appear to have turned off permissions for project maintainers to commit to your PR branch. Ordinarily I would have been happy to add these extra tweaks myself rather than go through a re-review; but in this case, I can't push to your branch. Regardless of whether it's intentional or accidental, it's not problem - it just means I can't do a quick tweak and merge. If nothing else, now you get the satisfaction of saying you did the *whole* patch :-)

**Comment by reefwing** [comment]
Thanks for the support and kind words. It was definitely not intentional to turn off project maintainers commit permission. In fact I don't know how I did that!

**Comment by reefwing** [comment]
@freakboy3742 - I've made the selected changes. For some reason I can't edit the PR to allow you to commit changes. I suspect I had the wrong box checked when creating the PR. I'll keep an eye out for this next time.

**Review by freakboy3742 — APPROVED** [review]
Looks great - Thanks for the contribution!

**Pull request merged**